### DISCUSSION TITLE
Show only allowed scopes under each Auth Flow in the generated Swagger UI

### DISCUSSION THREAD

**Discussion by Abdelgha-4** [fastapi/fastapi#12090] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
oauth2_scheme = OAuth2(
    flows=OAuthFlows(
        implicit=OAuthFlowImplicit(
            authorizationUrl="",
            scopes={"scope0": "Common scope", "scope1": "Implicit only scope"},
        ),
        clientCredentials=OAuthFlowClientCredentials(
            tokenUrl="",
            scopes={"scope0": "Common scope", "scope2": "ClientCredentials only scope"},
        ),
    )
)

async def get_current_user(
    security_scopes: SecurityScopes, token: Annotated[str, Depends(oauth2_scheme)]
):
    ...

async def get_implicit_user_with_scope1(
    # I can't specify in the Swagger UI that endpoints depending on this method should allow scope1 only through implicit flow
    current_user: Annotated[User, Security(get_current_user, scopes=["scope1"])],
):
    if current_user.disabled:
        raise HTTPException(status_code=400, detail="Inactive user")
    return current_user

async def get_client_with_scope2(
    # Similarly I can't specify that, here, only scope2 with client flow should be allowed
    current_user: Annotated[User, Security(get_current_user, scopes=["scope2"])],
):
    if current_user.disabled:
        raise HTTPException(status_code=400, detail="Inactive user")
    return current_user
```

### Description

The generated Swagger by FastAPI when using `scopes` argument from `Security` assumes that all the scopes are allowed for all Auth Flows. However this may not be the case, for example an endpoint may be accessible via implicit flow with "scope1" while accessible via explicit flow with "scope2" and "scope1", or may be accessible only via implicit flow etc..

I suppose that this issue could've been resolved if the `scopes` argument accept a dict-like object that maps allowed auth flows to required scopes. Then FastAPI could use this information to display the scopes and auth flows correctly.

```python
# only scope2 and 3 through implicit flow
Security(get_current_user, scopes={"implicit": ["scope2", "scope3"]})

# scope1 with implicit flow, or both scopes 1 & 2 with client flow
Security(get_current_user, scopes={"implicit": ["scope1"], "clientCredentials": ["scope1", "scope2"]})
```

### Operating System

Windows

### FastAPI Version

0.111.0

### Pydantic Version

2.7.1

### Python Version

3.11

**Comment by yakubka** [comment]
Swagger UI renders the scopes list from the `securitySchemes` definition in the OpenAPI spec. FastAPI populates that list from whatever you pass to the `OAuth2` or `OAuth2AuthorizationCodeBearer` constructor. There is no automatic filtering per endpoint.

To show only the scopes that apply to a specific flow, define separate security schemes for each flow with their own scope lists.

```python
from fastapi.security import OAuth2AuthorizationCodeBearer

read_scheme = OAuth2AuthorizationCodeBearer(
    authorizationUrl="/auth",
    tokenUrl="/token",
    scopes={"read:items": "Read items"},
)

write_scheme = OAuth2AuthorizationCodeBearer(
    authorizationUrl="/auth",
    tokenUrl="/token",
    scopes={"write:items": "Write items"},
)
```

Then apply the appropriate scheme to each endpoint via the `Security` dependency. Swagger UI will show each scheme separately with its own scope list.

If you use a single scheme with many scopes, Swagger UI has no way to know which scopes are relevant to which endpoint, so it shows all of them. Splitting into multiple schemes is the only way to control what Swagger UI displays per flow.

**Comment by MhussainD4772** [comment]
Hello!! If I'm reading this right, your goal is to have Swagger's Authorize dialog show different scopes per flow  scope1 under the implicit flow, scope2 under client credentials rather than every scope showing up under every flow. That part is achievable today. it just doesn't go through a per-scope dict.

The reason the scopes={"implicit": [...], "clientCredentials": [...]} shape can't be generated in OpenAPI, the thing attached to an operation is a Security Requirement Object, which maps security-scheme name, list of scope strings. There's no flow dimension in  the flows live in the scheme definition, but the per-operation requirement can only name a scheme and list scopes. So there's nowhere in the output format to say this scope, but only for this flow.
The OpenAPI docs actually call out this exact situation and recommend the workaround: define a separate security scheme per flow, each with its own scopes. Adapted to your setup:

```
from fastapi import FastAPI, Security, Depends
from fastapi.security import OAuth2, SecurityScopes
from fastapi.openapi.models import (
    OAuthFlows, OAuthFlowImplicit, OAuthFlowClientCredentials,
)

app = FastAPI()

implicit_scheme = OAuth2(
    scheme_name="ImplicitFlow",
    flows=OAuthFlows(
        implicit=OAuthFlowImplicit(
            authorizationUrl="https://example.com/auth",
            scopes={"scope0": "Common scope", "scope1": "Implicit only scope"},
        )
    ),
    auto_error=False,
)

client_credentials_scheme = OAuth2(
    scheme_name="ClientCredentialsFlow",
    flows=OAuthFlows(
        clientCredentials=OAuthFlowClientCredentials(
            tokenUrl="https://example.com/token",
            scopes={"scope0": "Common scope", "scope2": "ClientCredentials only scope"},
        )
    ),
    auto_error=False,
)

def via_implicit(scopes: SecurityScopes, token: str = Depends(implicit_scheme)):
    return token

def via_client_credentials(scopes: SecurityScopes, token: str = Depends(client_credentials_scheme)):
    return token

@app.get("/items/")
def read_items(
    u1=Security(via_implicit, scopes=["scope1"]),
    u2=Security(via_client_credentials, scopes=["scope2"]),
):
    return {"ok": True}
```

This produces two separate Authorize sections in Swagger, each showing only its own flow's scopes which is the behavior you're after.

One caveat worth flagging: this controls the OpenAPI/Swagger representation. At the spec level the two requirements form an OR (either scheme satisfies the operation), which matches your goal. But actual auth enforcement whether a request needs one token or both, and what the scopes mean  still lives in your own token-validation logic; splitting the schemes doesn't change how anything is enforced at runtime.

Hope that helps!

**Reply by Abdelgha-4** [reply]
Hello, Thank you for your answer!
My request is just a bit more general: not just different scopes per flow, but even per Endpoint. Say all my endpoints use client_credentials_scheme but Endpoint1 uses scope1 and Endpoint2 uses scope2, it would be then unreasonable to create as many schemes as endpoints, I'm hoping that the `scopes` argument passed to `Security` would actually change what scopes are shown required on each endpoint.

**Reply by initial-d** [reply]
For the per-endpoint case, there are two different Swagger UI surfaces that are easy to mix up:

1. **The operation security requirement**

   This is per endpoint. FastAPI already emits only the scopes you pass to `Security(..., scopes=[...])` for that operation.

   Conceptually:

   ```json
   "security": [
     { "OAuth2PasswordBearer": ["scope1"] }
   ]
   ```

   for endpoint 1, and:

   ```json
   "security": [
     { "OAuth2PasswordBearer": ["scope2"] }
   ]
   ```

   for endpoint 2.

2. **The Swagger UI Authorize dialog**

   This is not per endpoint. It is built from `components.securitySchemes.<scheme>.flows.<flow>.scopes`, which is the full list of scopes supported by that OAuth scheme. Swagger UI treats authorization as a global/session-level action, so it shows all scopes declared by the scheme, not only scopes needed by the endpoint you are currently looking at.

So if the goal is “the generated OpenAPI operation should require only `scope1` for endpoint 1 and only `scope2` for endpoint 2”, FastAPI should already be able to express that with:

```python
@app.get("/endpoint-1")
def endpoint_1(user=Security(get_current_user, scopes=["scope1"])):
    ...

@app.get("/endpoint-2")
def endpoint_2(user=Security(get_current_user, scopes=["scope2"])):
    ...
```

But if the goal is “when I click Authorize, only show the scopes relevant to the endpoint I care about right now”, that is a Swagger UI behavior limitation. The modal is scheme-level, not operation-level. FastAPI cannot encode a per-endpoint filtered authorize modal in OpenAPI without inventing many separate security schemes.

That is why separate schemes are the only spec-compatible workaround when you need the modal itself to be split. It is not ideal for many endpoints, but it matches the OpenAPI model: operations reference schemes + required scope lists; schemes define the global available scope catalog.

**Comment by YuriiMotov** [comment]
I think the way to go is creating 2 separate security schemes with `auto_error=Fasle`:

```py
oauth2_implicit_scheme = OAuth2(
    flows=OAuthFlows(
        clientCredentials=OAuthFlowClientCredentials(
            tokenUrl="",
            scopes={"scope0": "Common scope", "scope2": "ClientCredentials only scope"},
        ),
    ),
    auto_error=False,
)

oauth2_client_cred_scheme = OAuth2(
    flows=OAuthFlows(
        clientCredentials=OAuthFlowClientCredentials(
            tokenUrl="",
            scopes={"scope0": "Common scope", "scope2": "ClientCredentials only scope"},
        ),
    ),
    auto_error=False,
)
```

And then you can create a helper function that would construct the combination you need for each endpoint
